guides/hack/11-built-in-types/35-enum-class.md

```diff
@@ -85,30 +101,68 @@ enum class EC : int {
 
+But if we look at the enum class `EC::A` its type is `HH\MemberOf<EC, int>`. We know that it's declared within the enum class `EC`, with type `int`.
 
-Let's have a look back at `Names::World`, its type is `HH\MemberOf<Names, HasName>`, and `Names::Bar` has type `HH\MemberOf<Names, ConstName>`.
+## Extending an Existing enum class (Inheritance)
 
-The addition of this type layer is here to give more information about the enumeration, namely its exact type
-(`HasName` vs `IHasName`) and from which enum class it comes from (`Names`).
-`HH\MemberOf` is designed to allow a transparent access to the underlying value: `HH\MemberOf<Names, HasName> <: HasName`.
-This means that this layer can be ignored if one doesn’t need the additional information, and `Names::Hello` can be used directly as an instance of the `HasName` class:
+Enum classes can be composed together, as long as they implement the same base type:
 
-```EnumClassIntro.names.code0.hack no-auto-output
-function expect_name(HasName $x) : void {}
-// Names::Hello has type HH\MemberOf<Names, HasName>
+```EnumClassBox.extends0.hack no-auto-output
+enum class EBase : IBox {
+  Box<int> Age = new Box(42);
+}
 
-function test0(): void {
-  expect_name(Names::Hello); // ok !
+enum class EExtend : IBox extends EBase {
+  Box<string> Color = new Box('red');
 }
 ```.ini
 hhvm.hack.lang.enable_enum_classes=1
 ```
 
-## Defining function that expects enum class
+In this example, `EExtend` inherits `Age` from `EBase`, which means that `EExtend::Age` is defined.
+
+As with ordinary class extension, using the `extends` keyword will create a subtype relation between the enums: `EExtend <: EBase`.
+Enum classes support multiple inheritance as long as there is no ambiguity in the names of the constants, and that each enum class uses the same base type:
+
+```EnumClassBox.extends1.hack no-auto-output
+enum class E : IBox {
+  Box<int> Age = new Box(42);
+}
+
+enum class F : IBox {
+  Box<string> Name = new Box('foo');
+}
+
+enum class X : IBox extends E, F { } // ok, no ambiguity
+
+
+enum class E0 : IBox extends E {
+  Box<int> Color = new Box(0);
+}
+
+enum class E1 : IBox extends E {
+  Box<string> Color = new Box('red');
+}
+
+// enum class Y : IBox extends E0, E1 { }
+// type error, Y::Color is declared twice, in E0 and in E1
+// only he name is use for ambiguity
+```.ini
+hhvm.hack.lang.enable_enum_classes=1
+```
+
+### Control over inheritance
+
+Enum classes support the `__Sealed` attribute, just like normal classes, which allows for tighter control over extension mechanics.
+However enum classes do not yet support the `final` keyword.
+
+## Defining a Function that expects an enum class
 
-As previously explained, it is completely fine to use enum class constants as an instance of their declared type:
+It's completely fine to use enum class constants as an instance of their declared type.
```

Comment:
Like enums, the values of an enum class are subtypes of the enum itself - e.g. `MyEnumClass::MY_VALUE` is a valid parameter for a function expecting a `MyEnumClass`.

Reply:
No they are not. There is no value of type `MyEnumClass`. For example `E::Age` doesn't have type `E`. It has type `HH\MemberOf<E, Box<int>>` which is a subtype of `Box<int>`. You can consider `E` as a namespace rather than a type.

Reply:
"The declared type" here would be `Box<int>`, not `E`

Reply:
Whoops; I thought this would work, but perhaps it should be given as an example and explicitly explain why it is invalid?

```Hack
enum class Foo: string {
  string BAR = 'BAZ';
}

function do_stuff(Foo $value): void {
  var_dump($value);
}

<<__EntryPoint>>
function main(): void {
  do_stuff(Foo::BAR);
}
```

We shouldn't use the word namespace though unless we're very clear that `MyEnumClass\foo` is invalid.

Reply:
An "invalid" example might do the trick. Going into the details why this doesn't work involves explaining lots of subtyping bits that will clutter the doc I'm afraid